core/capabilities/launcher.go

```diff
@@ -0,0 +1,525 @@
+	idsToDONs := map[string]capabilities.DON{}
+	for _, d := range remoteWorkflowDONs {
+		idsToDONs[fmt.Sprint(d.Id)] = *toDONInfo(d)
+	}
+
+	for _, c := range don.CapabilityConfigurations {
+		capability, ok := state.IDsToCapabilities[c.CapabilityId]
+		if !ok {
+			return fmt.Errorf("could not find capability matching id %s", c.CapabilityId)
+		}
+
+		switch toCapabilityType(capability.CapabilityType) {
+		case capabilities.CapabilityTypeTrigger:
+			newTriggerPublisher := func(capability capabilities.BaseCapability, info capabilities.CapabilityInfo) (receiverService, error) {
+				cfg := &remotetypes.RemoteTriggerConfig{}
+				cfg.ApplyDefaults()
+				err := proto.Unmarshal(c.Config, cfg)
+				if err != nil {
+					return nil, err
+				}
+				publisher := remote.NewTriggerPublisher(
+					cfg,
+					capability.(capabilities.TriggerCapability),
+					info,
+					*toDONInfo(don),
+					idsToDONs,
+					w.dispatcher,
+					w.lggr,
+				)
+				return publisher, nil
+			}
+
+			err := w.addReceiver(ctx, capability, don, newTriggerPublisher)
+			if err != nil {
+				return fmt.Errorf("failed to add server-side receiver: %w", err)
+			}
+		case capabilities.CapabilityTypeAction:
+			w.lggr.Warn("no remote client configured for capability type action, skipping configuration")
+		case capabilities.CapabilityTypeConsensus:
+			w.lggr.Warn("no remote client configured for capability type consensus, skipping configuration")
+		case capabilities.CapabilityTypeTarget:
+			newTargetServer := func(capability capabilities.BaseCapability, info capabilities.CapabilityInfo) (receiverService, error) {
+				return target.NewServer(
+					myPeerID,
+					capability.(capabilities.TargetCapability),
+					info,
+					*toDONInfo(don),
+					idsToDONs,
+					w.dispatcher,
+					defaultTargetRequestTimeout,
+					w.lggr,
+				), nil
+			}
+
+			err := w.addReceiver(ctx, capability, don, newTargetServer)
+			if err != nil {
+				return fmt.Errorf("failed to add server-side receiver: %w", err)
+			}
+		default:
+			w.lggr.Warnf("unknown capability type, skipping configuration: %+v", capability)
+		}
+	}
+	return nil
+}
+
+type receiverService interface {
+	services.Service
+	remotetypes.Receiver
+}
+
+func (w *launcher) addReceiver(ctx context.Context, capability kcr.CapabilitiesRegistryCapabilityInfo, don kcr.CapabilitiesRegistryDONInfo, newReceiverFn func(capability capabilities.BaseCapability, info capabilities.CapabilityInfo) (receiverService, error)) error {
+	fullCapID := fmt.Sprintf("%s@%s", capability.LabelledName, capability.Version)
+	info, err := capabilities.NewRemoteCapabilityInfo(
+		fullCapID,
+		toCapabilityType(capability.CapabilityType),
+		fmt.Sprintf("Remote Capability for %s", fullCapID),
+		toDONInfo(don),
+	)
+	if err != nil {
+		return err
+	}
+	underlying, err := w.registry.Get(ctx, fullCapID)
```

Comment:
add context to these errors?
e.g.:

underlying, err := w.registry.Get(ctx, fullCapID)
	if err != nil {
		return fmt.Errorf("failed to get underlying capability from registry: %w", err)
	}